crates/extensions/tedge_mqtt_ext/src/lib.rs

```diff
@@ -120,22 +209,60 @@ impl FromPeers {
     async fn relay_messages_to(
         &mut self,
         outgoing_mqtt: &mut mpsc::UnboundedSender<MqttMessage>,
+        client: mqtt_channel::AsyncClient,
+        subscriptions: Arc<Mutex<TopicFilter>>,
     ) -> Result<(), RuntimeError> {
         while let Ok(Some(message)) = self.try_recv().await {
-            tracing::debug!(target: "MQTT pub", "{message}");
-            SinkExt::send(outgoing_mqtt, message)
-                .await
-                .map_err(Box::new)?;
+            match message {
+                PublishOrSubscribe::Publish(message) => {
+                    tracing::debug!(target: "MQTT pub", "{message}");
+                    SinkExt::send(outgoing_mqtt, message)
+                        .await
+                        .map_err(Box::new)?;
+                }
+                PublishOrSubscribe::Subscribe(mut request) => {
+                    let overlap = {
+                        let mut subscriptions = subscriptions.lock().unwrap();
+                        subscriptions.add_all(request.0.clone());
+                        subscriptions.remove_overlapping_patterns()
+                    };
+                    let mut to_unsub = TopicFilter::empty();
+                    for o in overlap {
+                        if request.0.remove(&o).is_none() {
+                            to_unsub.add_unchecked(&o);
+                        }
+                    }
+                    let to_sub = request.0;
+                    let to_sub = to_sub.filters();
+                    if !to_sub.is_empty() {
+                        info!("Updating MQTT subscription to include {to_sub:?}");
+                        client.subscribe_many(to_sub).await.unwrap();
```

Comment:
> NB This doesn't currently work since dynamic subscriptions aren't hooked up fully. tedge_mqtt_ext needs some further modifications to support sending the received messages on newly subscribed topics to the actor that requested the subscription

Okay,  I understand this comment now. The MQTT subscription is correctly updated *but* the messages are not forwarded by this actor which `ToPeers::peer_senders` subscriptions list has not been updated.

Indeed, not easy to update  from the `FromPeers` event loop this `peer_senders` list which is owned by the `ToPeers` event loop. A new channel, would be not enough as these peer `Sender`s are anonymous. Which one would have to be updated?

Reply:
Here is a proposal to fix that: https://github.com/thin-edge/thin-edge.io/pull/3626/commits/43932cf2250b37807ce77e3408ad8c35029e240c